src/lj_jitlog.c

```diff
@@ -0,0 +1,196 @@
+#define LUA_CORE
+
+#include "lj_jit.h"
+#include "lj_vm.h"
+#include "lj_lib.h"
+#include "lj_trace.h"
+#include "lj_tab.h"
+#include "lj_gc.h"
+#include "lj_buf.h"
+#include "lj_vmevent.h"
+#include "lj_debug.h"
+#include "luajit.h"
+#include "lauxlib.h"
+
+#include "lj_jitlog_def.h"
+#include "lj_jitlog_writers.h"
+
+typedef struct JITLogUserContext {
+  void *userdata;
+} JITLogUserContext;
+
+typedef struct JITLogState {
+  SBuf eventbuf; /* Must be first so loggers can reference it just by casting the G(L)->vmevent_data pointer */
+  JITLogUserContext user;
+  global_State *g;
+} JITLogState;
+
+
+static void free_context(JITLogState *context);
+
+static void jitlog_callback(void *contextptr, lua_State *L, int eventid, void *eventdata)
+{
+  VMEvent2 event = (VMEvent2)eventid;
+  JITLogState *context = contextptr;
+#if LJ_HASJIT
+  jit_State *J = L2J(L);
+#else
+  jit_State *J = NULL;
+#endif
+
+  switch (event) {
+    case VMEVENT_DETACH:
+      free_context(context);
+      break;
+    case VMEVENT_STATE_CLOSING:
+      if (G(L)->vmevent_cb == jitlog_callback) {
+        luaJIT_vmevent_sethook(L, NULL, NULL);
+      }
+      free_context(context);
+      break;
+    default:
+      break;
+  }
+}
+
+#if LJ_TARGET_X86ORX64
+
+static int getcpumodel(char *model)
+{
+  lj_vm_cpuid(0x80000002u, (uint32_t*)(model));
+  lj_vm_cpuid(0x80000003u, (uint32_t*)(model + 16));
+  lj_vm_cpuid(0x80000004u, (uint32_t*)(model + 32));
+  return (int)strnlen((char*)model, 12 * 4);
+}
+
+#else
+
+static int getcpumodel(char *model)
+{
+  strcpy(model, "unknown");
+  return (int)strlen("unknown");
+}
+
+#endif
+
+static int bufwrite_strlist(SBuf *sb, const char** list, int limit)
+{
+  const char** pos = list;
+  int count = 0;
+  for (; *pos != NULL && (limit == -1 || count < limit); pos++, count++) {
+    const char *s = *pos;
+    lj_buf_putmem(sb, s, (MSize)strlen(s)+1);
+  }
+  return count;
+}
+
+static void write_header(JITLogState *context)
+{
+  lua_State *L = mainthread(context->g);
+  char cpumodel[64] = {0};
+  int model_length = getcpumodel(cpumodel);
+  SBuf sb;
+  lj_buf_init(L, &sb);
+  bufwrite_strlist(&sb, msgnames, MSGID_MAX);
+  log_header(context->g, 1, 0, sizeof(MSG_header), msgsizes, MSGID_MAX, sbufB(&sb), sbuflen(&sb), cpumodel, model_length, LJ_OS_NAME, (uintptr_t)G2GG(context->g));
+  lj_buf_free(context->g, &sb);
+}
+
+int jitlog_isrunning(lua_State *L)
+{
+  void* current_context = NULL;
+  luaJIT_vmevent_callback cb = luaJIT_vmevent_gethook(L, (void**)&current_context);
+  return cb == jitlog_callback;
+}
+
+LUA_API JITLogUserContext* jitlog_start(lua_State *L)
+{
+  JITLogState *context ;
+  lua_assert(!jitlog_isrunning(L));
+
+  context = malloc(sizeof(JITLogState));
+  memset(context, 0 , sizeof(JITLogState));
+  context->g = G(L);
+
+  MSize total = G(L)->gc.total;
+  SBuf *sb = &context->eventbuf;
+  lj_buf_init(L, sb);
+  lj_buf_more(sb, 1024*1024*100);
+  /* Don't distort the gc rate by our buffer size */
```

Comment:
Can we try rephrasing this? I'm not sure I understand it.

Reply:
Its kind of a dirty hack work around making the gc getting triggered less for tests because its based on heap growth between successive GC's

Reply:
Can you expand the comment to make this clearer please? Thanks!

Reply:
updated the comment